### PR TITLE
(WIP)  Pad each batch, not the whole dataset

### PR DESCRIPTION
**Pull request by sshleifer**
This is a placeholder PR so that others don't waste effort implement padding on the batch level from scratch, and should only be merged/reviewed after [https://github.com/huggingface/transfer-learning-conv-ai/pull/29](url) is merged.
(It branches off that branch).

Previously, each sequence was padded to the length of the longest sequence in the *dataset*.
In this PR, each *batch* is padded to the length of the longest sequence in the batch. This results in a 30% speedup with negligible impact on metrics.

### Code Changes
* `ChatDataset` yields example dicts like `{'input_ids': [[hist + cand1], ..[hist +cand_n]],}` for the `PADDED_INPUTS` and `mc_token_ids` and `mc_labels` in the same format as previously.
* `ChatDataset().collate_fn(examples: list) ` turns a list of example dicts into the list of 5 tensors by batching them and padding them
* As a result, `get_dataloaders` does much less
* There is a data format change to the part of the process where we make lists of examples to facilitate this.
* `convai_evaluation.py` still calls the old `pad_dataset`

### 1 Epoch Sanity Check

**Before Change: 85 minutes**
Validation: {'accuracy': 0.7483655941545956,
 'average_accuracy': 0.7483655941545956,
 'average_nll': 2.6815188920676687,
 'average_ppl': 14.607263311061963,
 'nll': 2.6815188920676687}

**After Change: 60 minutes**
Validation: {'accuracy': 0.7466991411357519,
 'average_accuracy': 0.7466991411357519,
 'average_nll': 2.6821035040007972,
 'average_ppl': 14.615805388160778,
 'nll': 2.6821035040007972}

**Command:**
```bash
python train.py --model_checkpoint openai-gpt --dataset_cache dataset_cache --fp16 O1 --n_epochs 1 --train_batch_size 4
```

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -72,11 +73,63 @@ def build_input_from_segments(persona, history, reply, tokenizer, lm_labels=Fals
     return instance, sequence  # TODO: second arg is never used, delete it
 
 
+def pad_and_tensorize(batch_dict, padding):
```
Comment (sshleifer) [inline train.py]: this and `ChatDataset` should be easy to unit test

File: train.py [hunk 2]
```diff
@@ -72,11 +73,63 @@ def build_input_from_segments(persona, history, reply, tokenizer, lm_labels=Fals
...
     personachat = get_dataset(tokenizer, args.dataset_path, args.dataset_cache)
 
     logger.info("Build inputs and labels")
+    datasets: dict = make_data_lists(args, personachat, tokenizer)
+    pad_id = tokenizer.convert_tokens_to_ids(SPECIAL_TOKENS[-1])
+    train_dataset = ChatDataset(datasets['train'], pad_id)
+    valid_dataset = ChatDataset(datasets['valid'], pad_id)
+
+    logger.info("Build train and validation dataloaders")
+    train_sampler = torch.utils.data.distributed.DistributedSampler(train_dataset) if args.distributed else None
```
Comment (sshleifer) [inline train.py]: (maybe) put this in `ChatDataset.to_loader(self, args, shuffle) -> sampler, loader`

Comment (sshleifer) [reply train.py]: at some point might also want to document which tensors are 3D

File: train.py [hunk 3]
```diff
@@ -86,36 +139,20 @@ def get_data_loaders(args, tokenizer):
...
                     for j, candidate in enumerate(utterance["candidates"][-num_candidates:]):
-                        lm_labels = bool(j == num_candidates-1)
-                        instance, _ = build_input_from_segments(persona, history, candidate, tokenizer, lm_labels)
+                        lm_labels = bool(j == num_candidates - 1)
+                        instance, _ = build_input_from_segments(persona, history, candidate,
+                                                                tokenizer, lm_labels)
                         for input_name, input_array in instance.items():
-                            datasets[dataset_name][input_name].append(input_array)
+                            candidate_instances[input_name].append(input_array)
+                    for k in candidate_instances.keys():
```
Comment (sshleifer) [inline train.py]: `.items()` will save some chars

File: train.py [hunk 4]
```diff
@@ -86,36 +139,20 @@ def get_data_loaders(args, tokenizer):
             persona = dialog["personality"].copy()
             for _ in range(args.personality_permutations):
                 for utterance in dialog["utterances"]:
-                    history = utterance["history"][-(2*args.max_history+1):]
+                    candidate_instances = defaultdict(list)
+                    history = utterance["history"][-(2 * args.max_history + 1):]
                     for j, candidate in enumerate(utterance["candidates"][-num_candidates:]):
-                        lm_labels = bool(j == num_candidates-1)
-                        instance, _ = build_input_from_segments(persona, history, candidate, tokenizer, lm_labels)
+                        lm_labels = bool(j == num_candidates - 1)
```
Comment (sshleifer) [inline train.py]: better varname?
